### PR TITLE
Initial API draft

### PR DESCRIPTION
**Pull request by rgbkrk**
This is a quick mock up of the initial API as discussed in the [enhancement proposal for the binder API](https://github.com/rgbkrk/enhancement-proposals/blob/b8e279bd84ffa6ba0bc41ac0b48d6c35dfe359d1/binder-api/README.md#register-templates). It's still WIP, but I'll keep iterating. Questions are coming up while implementing, which I'll write here or stick in as TODOs for the time being.

Remaining to do:

* [X] `GET /templates`
* [X] `POST /templates/`
* [X] `GET /templates/{templateName}`
* [X] `PUT /templates/{templateName}`
* [X] Require Authorization for POST
* [X] Require Authorization for PUT

### CODE REVIEW

**Comment by rgbkrk** [comment]
Direct `curl`ing (with some `json` beautification):

```
$ curl -s 127.0.0.1:8080/templates | json[]
$ curl -s -H "Content-Type: application/json" -d '{"image-name": "jupyter/notebook"}' 127.0.0.1:8080/templates | json
{
  "id": 0,
  "image-name": "jupyter/notebook",
  "command": "",
  "limits": {},
  "time-created": "2015-10-03T06:07:19.629914954Z",
  "time-modified": "2015-10-03T06:07:19.629914954Z"
}
$ curl -s -H "Content-Type: application/json" -d '{"image-name": "ceteri/jem"}' 127.0.0.1:8080/templates | json
{
  "id": 1,
  "image-name": "ceteri/jem",
  "command": "",
  "limits": {},
  "time-created": "2015-10-03T06:07:39.789190004Z",
  "time-modified": "2015-10-03T06:07:39.789190004Z"
}
$ curl -s 127.0.0.1:8080/templates | json[
  {
    "id": 0,
    "image-name": "jupyter/notebook",
    "command": "",
    "limits": {},
    "time-created": "2015-10-03T06:07:19.629914954Z",
    "time-modified": "2015-10-03T06:07:19.629914954Z"
  },
  {
    "id": 1,
    "image-name": "ceteri/jem",
    "command": "",
    "limits": {},
    "time-created": "2015-10-03T06:07:39.789190004Z",
    "time-modified": "2015-10-03T06:07:39.789190004Z"
  }
]
```

Associated logs:

```
$ ./binder-registry
2015/10/03 02:07:18 GET	/templates	TemplateIndex	178.548µs
2015/10/03 02:07:19 POST	/templates	TemplateCreate	113.576µs
2015/10/03 02:07:39 POST	/templates	TemplateCreate	49.601µs
2015/10/03 02:07:43 GET	/templates	TemplateIndex	63.629µs
2015/10/03 02:08:37 GET	/templates	TemplateIndex	41.448µs
```

**Comment by rgbkrk** [comment]
Some HTTPie input and output for fun (plus headers!):

screenshot 2015-10-03 02 12 52

**Comment by rgbkrk** [comment]
screenshot 2015-10-03 12 48 28

**Comment by rgbkrk** [comment]
Since we're working in go here, the namespace would be much better if it wasn't binder-prefixed. Currently imports look like:

```
import (
  registry "github.com/binder-project/binder-registry"
)
```

and it would be cleaner (since we're org level) to make the repo name be:

```
import (
  registry "github.com/binder-project/registry"
)
```

It's still the binder registry as a project, but the namespace is less deep.

**Comment by freeman-lab** [comment]
@rgbkrk Hmm not fundamentally opposed, though, we may be heading into node / npm land for at least some of the other modules :stuck_out_tongue:  see https://github.com/binder-project/binder-cli/pull/1

In which case, the `binder` prefix on repos makes more sense. Of course, package names need not match repo names, so the npm module names could have the `binder` prefix and not the repos themselves, if that's better for go. But consistency of repo names seems desirable.

**Comment by rgbkrk** [comment]
@freeman-lab I'm all about node for that layer, it's A-OK. That naming makes sense for both node and python (package name prefixed, even if repo is named something else).

File: registry/template.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"time"
+)
+
+// ContainerLimits is the specific set of limits we have for containers
+// QUESTION: Should this come from Docker itself?
+type ContainerLimits struct {
+	Memory    int64 `json:"memory,omitempty"`
+	CPUShares int64 `json:"cpu_shares,omitempty"`
```
Comment (captainsafia) [inline registry/template.go]: I'm being super picky, but I feel as though this should be delimited by "-" as opposed to "_" as in all the other key names.

Comment (rgbkrk) [inline registry/template.go]: Hmmm, we'll want to make sure we address that in the spec too.

Comment (captainsafia) [inline registry/template.go]: Done!

File: registry/apierrors.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+// UnavailableTemplateError is an APIErrorResponse with a boilerplate unavailable template message
+var UnavailableTemplateError = APIErrorResponse{Message: "Template unavailable"}
+
+// ExistingTemplateError is an APIErrorResponse with a boilerplate message for when a template already exists
+var ExistingTemplateError = APIErrorResponse{Message: "Template already exists"}
+
+// DontPanicError is an APIErrorResponse when there's a mostly catastrophic error
+var DontPanicError = APIErrorResponse{Message: "Internal Server Error. Don't Panic. We will."}
```
Comment (captainsafia) [inline registry/apierrors.go]: This is my favorite error message ever. I would not be disappointed to see this error message.

File: registry/template.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+// Template defines an image to be run along with its configuration
+type Template struct {
```
Comment (captainsafia) [inline registry/template.go]: So `Template` is akin to what I call `Environment` in the API spec?

File: registry/template.go (hunk 1 above)
Comment (rgbkrk) [inline registry/template.go]: The only question that is open to me is of this should just match the Docker API for container confit.

**Comment by rgbkrk** [comment]
This is pretty well wrapped up, including tests. It's not 100% covered (not all failure cases tested, especially those where the backing store responds with completely invalid data), but it's certainly working on all the common cases.

**Comment by captainsafia** [comment]
giphy-7

Would you mind opening up some issues for things that you would like fixed or added so I can take a look at them on Friday?

**Comment by freeman-lab** [comment]
great idea =) @andrewosh is close to having a draft of `binder-build`, and i'm likewise close on a draft of `binder-cli`, so can try to open issues there too!

**Pull request merged**